src/weak_memory.rs

```diff
@@ -0,0 +1,271 @@
+use std::{
+    cell::{Ref, RefCell, RefMut},
+    collections::VecDeque,
+};
+
+use rustc_const_eval::interpret::{AllocRange, InterpResult, ScalarMaybeUninit};
+use rustc_data_structures::fx::FxHashMap;
+use rustc_target::abi::Size;
+
+use crate::{
+    data_race::{GlobalState, ThreadClockSet},
+    RangeMap, Tag, VClock, VTimestamp, VectorIdx,
+};
+pub type AllocExtra = StoreBufferAlloc;
+#[derive(Debug, Clone)]
+pub struct StoreBufferAlloc {
+    /// Store buffer of each atomic object in this allocation
+    store_buffer: RefCell<RangeMap<StoreBuffer>>,
+}
+
+impl StoreBufferAlloc {
+    pub fn new_allocation(len: Size) -> Self {
+        Self { store_buffer: RefCell::new(RangeMap::new(len, StoreBuffer::default())) }
+    }
+    /// Gets a store buffer associated with an atomic object in this allocation
+    fn get_store_buffer(&self, range: AllocRange) -> Ref<'_, StoreBuffer> {
+        Ref::map(self.store_buffer.borrow(), |range_map| {
+            let (.., store_buffer) = range_map.iter(range.start, range.size).next().unwrap();
+            store_buffer
```

Comment:
This only gets the store buffer for the *first* fragment of that range. How does that ever make any sense?

You need to iterate over the range and do the same action on all elements you get. You cannot rely on how large the chunks are that the range_map gives you.

Reply:
Oh, but that iteration will be highly nontrivial. This gets into the realm of mixed-size atomic accesses. When you do a 4-byte atomic access, each of these 4 bytes might have a *different* store buffer and you need to somehow handle that.

Or maybe you don't intend to support mixed-size accesses... that's okay but then at the least this code must be able to detect that situation and bail out (and tell users what they have to do to still run their code).

Reply:
Mixed size atomic accesses are not allowed under the C++ model: https://www.cl.cam.ac.uk/~pes20/popl17/mixed-size.pdf. [This discussion](https://reviews.llvm.org/D56534#1364204) on LLVM semantics also implies it's not supported by LLVM.

We should make Miri throw this as UB, but probably not here since the UB exists whether or not weak memory (or even data race detector) is enabled.

Reply:
I didn't know mixed-size accesses are ruled out in C++. I think in Rust we have to support at least some kinds of them -- basically, when they are not racing: turning `&mut i64` into `&mut [i32; 2]` should be safe, and we can use that to first do `AtomicI64` stuff on some memory and later do `AtomicI32` stuff on the same memory (or vice versa). That should work fine in Miri, I think.

I could live with ruling out different-sized accesses that actually race with each other. However, it is still never okay to use `RangeMap` and then just get the first element of the iterator -- that is a clear violation of the API contract of `RangeMap`. Maybe you don't want a `RangeMap` then, and just something that maps "the first byte of the atomic" to the store buffers.

Reply:
I will make something called `NonSplittingRangeMap` that maps ranges to StoreBuffers without the splitting going on, so it can identify whether an access is a perfectly overlapping access (offset and size are equal to an existing entry), and such accesses only return one StoreBuffer.

I strongly believe we should throw an UB when a non-perfectly overlapping atomic access happens. There's no Rust specific spec on its behaviour and will probably never be one unless we define it in LLVM LangRef. 

Of course, it's not difficult for Miri to provide a value when this happens, and we could also disable weak memory over all atomic objects the non perfectly overlapping access touches, this would still be consistent with the memory model; but we don't know about what sort of optimisation behaviours this UB will lead to since instruction reordering heavily relies on the defined semantics of the memory model. This is the "everything goes with UB" part

Reply:
I just came across @m-ou-se's wonderful write ups on various implementation of synchronisation primitives. Apparently Wine's implementation of Windows SRW locks mixes atomic accesses of different sizes! https://github.com/rust-lang/rust/issues/93740#issuecomment-1055672041

Mara - what do you think about this in Rust? Should we throw a UB when Miri detects non-perfectly overlapping (different address and/or different size) atomic accesses?

Reply:
> Mara - what do you think about this in Rust? Should we throw a UB when Miri detects non-perfectly overlapping (different address and/or different size) atomic accesses?

Yes, if they actually race.

The Intel x86 manual says:

> Software should access semaphores (shared memory used for signalling between multiple processors) using identical addresses and operand lengths. For example, if one processor accesses a semaphore using a word access, other processors should not access the semaphore using a byte access.

In practice it will probably all be fine as long as there's nothing crossing a cache line boundary, but until we have a very strong confirmation that there are situations where it's fine, we should consider all non-perfectly overlapping atomic races UB.

But as Ralf said, when there's no race it should be accepted. `Atomic*::get_mut` and `from_mut` should be safe, and so should transmuting `&mut i64` to `&mut [i32; 2]` be.

> I think in Rust we have to support at least some kinds of them -- basically, when they are not racing: turning `&mut i64` into `&mut [i32; 2]` should be safe, and we can use that to first do `AtomicI64` stuff on some memory and later do `AtomicI32` stuff on the same memory (or vice versa). That should work fine in Miri, I think.

Reply:
Sure. I've made some major changes on how multiple atomic objects in the same allocation are represented, using the new `AllocationMap` type. An atomic access with an offset and size now has 3 possible access types:
- The accessed range is fully empty, meaning this is the first atomic access on this range, we just create an empty `StoreBuffer`
- The accessed range perfectly overlaps with an existing `StoreBuffer`, we just use that
- The accessed range imperfectly overlaps with one or more existing `StoreBuffer`. We remove all the existing ones it touches and create a new one in the accessed range.

Reading from an empty `StoreBuffer` results in the latest value (in mo) in every byte being observed by the program, I think this is a sane behaviour

This does not interfere with the race detector as it already works on individual bytes.

Reply:
> I strongly believe we should throw an UB when a non-perfectly overlapping atomic access happens. There's no Rust specific spec on its behaviour and will probably never be one unless we define it in LLVM LangRef.

That is fine for me.

But your last post sounds like you went for a different strategy?

Reply:
> But your last post sounds like you went for a different strategy?

I did since both you and Mara thinks it should be accepted as long as there is no race.

Though I might have misunderstood what you meant by no race here. Two atomic accesses by definition cannot cause data race. If we were to call two unsynchronised atomic accesses data race because they are imperfectly overlapping we'd be changing that definition, and we need to make some changes to the race detector. Right now it only detect data race when at least one access is non-atomic.

In any case, Wine's mixed 16 and 32 bit atomic accesses on the same location clearly are not synchronised: https://github.com/wine-mirror/wine/blob/3ded60bd1654dc689d24a23305f4a93acce3a6f2/dlls/ntdll/sync.c#L543-L566

Reply:
> When two atomic accesses are in a race condition and they imperfectly overlap

I tried to implement this and I found it quite tricky

(whenever I say unordered I mean "are not related by happens-before")

1. Say we do a 16-bit read on a location that's been written to twice with 32-bit values. Do we throw a UB when only one of them happens before our read but the other one doesn't?
```
  W32───┐
        │mo
        │
        ▼
       W32───┐
             │hb
             ▼
            R16
```

2. Do unordered read-reads count as "race condition"? And if the current read could race with not just the latest read, but ones prior as well, then we need to record the timestamp for all reads on a store element, which we currently aren't.
```
  R32───┐
        │execution order
        ▼
       R32───┐
             │execution order
             ▼
            R16
```

3. Unordered read-write should definitely be racy. But we run into the same issue of needing to record the timestamps on all reads
```
  R32─────┐
   │      │execution order
   │      ▼
   │hb   R32
   │      │
   ▼      │execution order
  W16◄────┘
```

5. Any implementation would also complicate the existing race detector, especially `MemoryCellClocks` and `AtomicMemoryCellClocks`.

I don't think it's worth the effort to implement these somewhat major changes to accommodate mixed size race detection. The previously linked paper made it clear that mixed size access is prohibited, and I hope this is also evident from our atomic object vs access discussion.

> In the C/C++11 model, programs that exhibit data races on non-atomics (or between non-atomic and atomic accesses) are deemed to have undefined behaviour, and the effective type rules of the ISO C standard prohibit mixed-size use of atomics. ([Shaked Flur et al.](https://www.cl.cam.ac.uk/~pes20/popl17/mixed-size.pdf))

Additionally, I don't quite understand what you mean by "race condition but not data race". I assume you call two accesses racy when they are not ordered by happens-before, but that is also the definition of data race. We _are_ effectively extending the definition of data race to pure atomic accesses.

Reply:
> I don't think it's worth the effort to implement these somewhat major changes to accommodate mixed size race detection. The previously linked paper made it clear that mixed size access is prohibited, and I hope this is also evident from our atomic object vs access discussion.

I am confused by this statement. If mixed size accesses are prohibited, then it is the job of tsan/Miri to detect them, isn't it? It is *because* they are prohibited that it is worth the effort to detect them -- the effort is even required if we want to detect all UB.

Of course, for now we could say that it is okay not to detect all UB, and document that. In that case I would still like to understand how far we are differing from an "ideal" Abstract Machine that detects all UB.

> Additionally, I don't quite understand what you mean by "race condition but not data race". I assume you call two accesses racy when they are not ordered by happens-before, but that is also the definition of data race. We are effectively extending the definition of data race to pure atomic accesses.

AFAIK this is standard terminology: a race condition is when two accesses (at least one of which is a write) on the same location happen without being ordered relative to each other. A data race is a race condition where at least one access is non-atomic.
For example, if you implement a spinlock with atomics, there is a race condition when two threads try to acquire the same lock concurrently. The threads are racing for being the first to get the lock. That's perfectly well-defined and all, but it is a concept worth studying in many situations (in particular, parallelism *without* any race conditions is perfectly deterministic).

Reply:
Sorry I meant it's difficult to detect whether a mixed size access is _racy_. It's a trivial code change to throw UB on mixed size atomic accesses without checking if it's ordered with anything else, this was my original intention.

The difficulty arises because the existing race detector was designed having in mind that racing pure atomic accesses are never UB so it's not worth checking if they are racy; but now after encountering a mixed size atomic access, we need to check if it's ordered with prior atomic accesses. We need relatively major changes to `data_race.rs` to achieve this.

I get the terminology now. I was kind of conflating "data race" with "UB", but yeah the UB here isn't a data race UB but something else.

Reply:
I don't quite understand what that means. Does this mean that using `from_mut` I can cause UB? In that case it should probably be `throw_unsup`, not `throw_ub`. And please make sure there is a test for this, if there isn't already. Something that starts with a `x: &mut i32`, uses `AtomicI32::from_mut`, does some atomic accesses, then drops the `AtomicI32` and transmutes `x` to `&mut [i8; 4]` and uses `AtomicI8::from_mut` and does some atomic accesses there. All in the same thread.

Reply:
Actually, even `throw_unsup` is a problem since this is a regression: that same code worked fine before this PR.